File: src/main/scala/resonant/engine/content/debug/BlockCreativeBuilder.java
package resonant.engine.content.debug;

import io.netty.buffer.ByteBuf;
import net.minecraft.block.Block;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.server.MinecraftServer;
import net.minecraft.world.World;
import net.minecraftforge.common.util.ForgeDirection;
import resonant.engine.ResonantEngine;
import resonant.lib.network.IPacketReceiver;
import resonant.lib.schematic.Schematic;
import resonant.lib.type.Pair;
import universalelectricity.api.UniversalElectricity;
import universalelectricity.core.transform.vector.Vector3;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map.Entry;

/**
 * Automatically set up structures to allow easy debugging in creative mode.
 */
public class BlockCreativeBuilder extends BlockRotatable implements IPacketReceiver
{
	public static final List<Schematic> REGISTRY = new ArrayList<Schematic>();

	public BlockCreativeBuilder(int id)
	{
		super(id, UniversalElectricity.machine);
		setCreativeTab(CreativeTabs.tabTools);
		rotationMask = Byte.parseByte("111111", 2);
	}

	public static int register(Schematic schematic)
	{
		REGISTRY.add(schematic);
		return REGISTRY.size() - 1;
	}

	/**
	 * Called when the block is right clicked by the player
	 */
	@Override
	public boolean onMachineActivated(World par1World, int x, int y, int z, EntityPlayer par5EntityPlayer, int side, float hitX, float hitY, float hitZ)
	{
		if (REGISTRY.size() > 0)
		{
			par5EntityPlayer.openGui(ResonantEngine.instance, -1, par1World, x, y, z);
			return true;
		}

		return false;
	}

	@Override
	public void onReceivePacket(ByteBuf data, EntityPlayer player, Object... extra)
	{
		World world = player.worldObj;

		if (!world.isRemote)
		{
			// Only allow operators.
			if (MinecraftServer.getServer().getConfigurationManager().isPlayerOpped(player.getCommandSenderName()))
			{
				try
				{
					int schematicID = data.readInt();
					int size = data.readInt();
					Vector3 position = new Vector3((Integer) extra[0], (Integer) extra[1], (Integer) extra[2]);

					if (size > 0)
					{
						HashMap<Vector3, Pair<Block, Integer>> map = REGISTRY.get(schematicID).getStructure(ForgeDirection.getOrientation(position.getBlockMetadata(world)), size);

						for (Entry<Vector3, Pair<Block, Integer>> entry : map.entrySet())
						{
							Vector3 placePos = entry.getKey().clone();
							placePos.$plus(position);
							placePos.setBlock(world, entry.getValue().left(), entry.getValue().right());
						}
					}
				}
				catch (Exception e)
				{
					e.printStackTrace();
				}
			}
		}
	}
}

File: src/main/scala/resonant/engine/content/debug/BlockCreativeBuilder.scala
package resonant.engine.content.debug

import java.util.{ArrayList, List}

import io.netty.buffer.ByteBuf
import net.minecraft.block.material.Material
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.server.MinecraftServer
import net.minecraft.world.World
import net.minecraftforge.common.util.ForgeDirection
import resonant.content.spatial.block.SpatialBlock
import resonant.engine.ResonantEngine
import resonant.lib.content.prefab.TRotatable
import resonant.lib.network.IPacketReceiver
import resonant.lib.schematic.Schematic
import universalelectricity.core.transform.vector.Vector3

import scala.collection.convert.wrapAll._

/**
 * Automatically set up structures to allow easy debugging in creative mode.
 */
object BlockCreativeBuilder
{
  def register(schematic: Schematic): Int =
  {
    registry.add(schematic)
    return registry.size - 1
  }

  final val registry: List[Schematic] = new ArrayList[Schematic]
}

class BlockCreativeBuilder extends SpatialBlock(Material.iron) with TRotatable with IPacketReceiver
{
  creativeTab = CreativeTabs.tabTools
  rotationMask = Integer.parseInt("111111", 2).toByte

  /**
   * Called when the block is right clicked by the player
   */
  override def activate(player: EntityPlayer, side: Int, hit: Vector3): Boolean =
  {
    if (BlockCreativeBuilder.registry.size > 0)
    {
      player.openGui(ResonantEngine.instance, -1, world, x, y, z)
      return true
    }
    return false
  }

  override def onReceivePacket(data: ByteBuf, player: EntityPlayer, extra: AnyRef*)
  {
    val world: World = player.worldObj
    if (!world.isRemote)
    {
      if (MinecraftServer.getServer.getConfigurationManager.isPlayerOpped(player.getCommandSenderName))
      {
        try
        {
          val schematicID: Int = data.readInt
          val size: Int = data.readInt
          val translation: Vector3 = new Vector3(extra(0).asInstanceOf[Double], extra(1).asInstanceOf[Double], extra(2).asInstanceOf[Double])
          if (size > 0)
          {
            val map = BlockCreativeBuilder.registry.get(schematicID).getStructure(ForgeDirection.getOrientation(translation.getBlockMetadata(world)), size)
            map.foreach(entry => (entry._1 + translation).setBlock(world, entry._2.left(), entry._2.right()))
          }
        }
        catch
          {
            case e: Exception =>
            {
              e.printStackTrace
            }
          }
      }
    }
  }
}